### ISSUE TITLE
Ошибка при повторном использовании переменной в шаблоне

### ISSUE THREAD

**Issue by psttf** [pomadchin/template-macro#5]
Макрос не компилируется, если в шаблоне повторяются переменные:

``` scala
case class Twice(first: Int, second: Int)
@template((i: Int) => Twice(i, i)) object SameTwice
```

Получаем ошибку:

```
[error] i is already defined as value i
[error]       @template((i: Int) => Twice(i, i)) object SameTwice
[error]        ^
```

Я сделал тест для проверки — [`TemplateSpec`](https://github.com/psttf/template-macro/blob/a4fa735901bc615412a658653eacd713a4531475/macros/src/test/scala/annotations/TemplateSpec.scala).

**Comment by psttf** [comment]
Гриша, а вот этот тест все еще не проходит

``` scala
      @template((i: Int) => Twice(i, i)) object SameTwice
      SameTwice unapply Twice(1, 2) mustEqual None
```

```
[error]    'Some(2)' is not equal to 'None'
```

**Comment by pomadchin** [comment]
Оу, забыл совсем про него.

**Comment by pomadchin** [comment]
Думаю, что не совсем корректное поведение: 

``` scala
@template((i: Int) => Twice(i, i)) object SameTwice
SameTwice unapply Twice(1, 1) //> Some(1)
SameTwice unapply Twice(1, 2) //> None
```

Думаю, что в этом случае необходимо или 

``` scala
SameTwice unapply Twice(1, 2) //> Some(1)
```

Или тогда придется генерировать функцию, которая будет сравнивать аргументы.

``` scala
a match {
  case T(i1, i2, i3) => if((i1 :: i2 :: i3 :: Nil).distinct.length == 1) Some(i1) else None
  case _ => None
}
```

**Comment by pomadchin** [comment]
https://github.com/pomadchin/template-macro/blob/82a1bacfc38cc9e38a2e5f8e79c8c302ad4bab12/macros/src/test/scala/annotations/TemplateSpec.scala#L37-L62

**Comment by psttf** [comment]
Вообще говоря, это ожидаемое поведение функции unapply: если `apply(X) = Y`, то `unapply(Y) = Some(X)`; а если `Z` не является экземпляром соответствующего класса, то `unapply(Z) = None`.

В моем примере `SameTwice` — это подкласс двойных значений (`Twice`), в которых оба значения одинаковы. Сооветственно `Twice(1, 1)` удовлетворяет этому определению, а `Twice(1, 2)` — нет, и `unapply` для него должно вернуть `None`.

**Comment by pomadchin** [comment]
Это верно с точки зрения типов. В момент компиляции, к сожалению, не возможно знать значение, поэтому для случая `Twice(1, 1)`, думаю нужен специальный тип, уже не относящийся к генерации. т.к. `unapply` будет ожидать все равно `case Twice(i1, i2) => Some(i1 -> i2)`, и с этим ничего не поделать. Да и функция `unapply Twice(1, 2)` в примере из спек, не имеет данных об использовании `apply` (и не может иметь), потому что должна уметь обработать как случай `Twice(a, a)`, так и случай `Twice(a, b)` (потому, что тип `Twice` такой).

**Comment by psttf** [comment]
да, для обработки подобных случаев скорее всего понадобится особый механизм: необходимо распознавать, что одна и та же переменная используется несколько раз и вставлять соответствующие проверки
